src/com/owncloud/android/ui/helpers/UriUploader.java

```diff
@@ -0,0 +1,188 @@
+                copyThenUpload(contentUris.toArray(new Uri[contentUris.size()]),
+                        contentRemotePaths.toArray(new String[contentRemotePaths.size()]));
+
+            } else if (schemeFileCounter == 0) {
+                mCode = UriUploaderResultCode.ERROR_NO_FILE_TO_UPLOAD;
+
+            }
+
+        } catch (SecurityException e) {
+            mCode = UriUploaderResultCode.ERROR_READ_PERMISSION_NOT_GRANTED;
+            Log_OC.e(TAG, "Permissions fail", e);
+
+        } catch (Exception e) {
+            mCode = UriUploaderResultCode.ERROR_UNKNOWN;
+            Log_OC.e(TAG, "Unexpted error", e);
+
+        } finally {
+            // Save the path to shared preferences; even if upload is not possible, user chose the folder
+            PreferenceManager.setLastUploadPath(mUploadPath, mActivity);
+        }
+        return mCode;
+    }
+
+    private String generateDiplayName() {
+        return mActivity.getString(R.string.common_unknown) +
+                "-" + DisplayUtils.unixTimeToHumanReadable(System.currentTimeMillis());
+    }
+
+    /**
+     * Requests the upload of a file in the local file system to {@link FileUploader} service.
+     *
+     * The original file will be left in its original location, and will not be duplicated.
+     * As a side effect, the user will see the file as not uploaded when accesses to the OC app.
+     * This is considered as acceptable, since when a file is shared from another app to OC,
+     * the usual workflow will go back to the original app.
+     *
+     * @param localPath     Absolute path in the local file system to the file to upload.
+     * @param remotePath    Absolute path in the current OC account to set to the uploaded file.
+     */
+    private void requestUpload(String localPath, String remotePath) {
+        FileUploader.UploadRequester requester = new FileUploader.UploadRequester();
+        requester.uploadNewFile(
+                mActivity,
+                mAccount,
+                localPath,
+                remotePath,
+                mBehaviour,
+                null,       // MIME type will be detected from file name
+                false,      // do not create parent folder if not existent
+                UploadFileOperation.CREATED_BY_USER
+        );
+    }
+
+    /**
+     *
+     * @param sourceUris        Array of content:// URIs to the files to upload
+     * @param remotePaths       Array of absolute paths to set to the uploaded files
+     */
+    private void copyThenUpload(Uri[] sourceUris, String[] remotePaths) {
+        if (mShowWaitingDialog) {
+            mActivity.showLoadingDialog(mActivity.getResources().
+                    getString(R.string.wait_for_tmp_copy_from_private_storage));
+        }
+
+        CopyAndUploadContentUrisTask copyTask = new CopyAndUploadContentUrisTask
+                ((CopyAndUploadContentUrisTask.OnCopyTmpFilesTaskListener)mActivity, mActivity);
```

Comment:
The casting of the first parameter in the last minute is slightly dangerous. Let's avoid it adding a separated parameter in the constructor to receive an ```OnCopyTmpFilesTaskListener``` instance.

Besides, NULL should be a valid value for it. ```FileDisplayActivity``` is doing nothing in ```onTmpFilesCopied(...)```, could pass a NULL value for the listener and wouldn't need to implement ```OnCopyTmpFilesTaskListener```.